library/hls/src/main/java/com/google/android/exoplayer2/source/hls/HlsSampleStreamWrapper.java

```diff
@@ -1337,26 +1340,57 @@ private static boolean formatsMatch(Format manifestFormat, Format sampleFormat)
     return true;
   }
 
-  private static DummyTrackOutput createDummyTrackOutput(int id, int type) {
-    Log.w(TAG, "Unmapped track with id " + id + " of type " + type);
-    return new DummyTrackOutput();
-  }
+  /**
+   * Extend base SampleQueue to add HLS specific processing of the samples, including:
+   *
+   * <ul>
+   *   <li>segment time boundary checks on timestamps of committed samples</li>
+   *   <li>cleaning the {@link Format#metadata} to avoid excessive format changes</li>
+   * </ul>
+   *
+   * The timestamp check verifies that the adjusted sample time (via {@link TimestampAdjuster}) does not
+   * fall outside of a set percentage ({@link #MAX_TIMESTAMP_DEVIATION_PERCENTAGE}) of the time
+   * boundaries of the segment as expressed by the segment duration ((@link HlsMediaChunk#endTimeUs} -
+   * {@link HlsMediaChunk#startTimeUs}).  This is loosely mandated by the Pantos spec and checked by
+   * Apple's mediastreamvalidator.
+   *
+   */
+  private static final class HlsSampleQueue extends SampleQueue {
 
-  private static final class FormatAdjustingSampleQueue extends SampleQueue {
+    private static final String TAG = "HlsSampleQueue";
+
+    /**
+     * largest timestamp deviation from the segment time bounds expressed as a percentage of
+     * the segment duration.
+     */
+    private static double MAX_TIMESTAMP_DEVIATION_PERCENTAGE = 0.50;
+
+    private long lowestTimeUs = C.TIME_UNSET;
+    private long highestTimeUs = C.TIME_UNSET;
+
+    @Nullable private HlsMediaChunk chunk;
+    private long lastValidTimeUs;
 
     private final Map<String, DrmInitData> overridingDrmInitData;
     @Nullable private DrmInitData drmInitData;
 
-    public FormatAdjustingSampleQueue(
-        Allocator allocator,
-        DrmSessionManager<?> drmSessionManager,
-        MediaSourceEventDispatcher eventDispatcher,
-        Map<String, DrmInitData> overridingDrmInitData) {
+    private HlsSampleQueue(Allocator allocator,
+            DrmSessionManager<?> drmSessionManager,
+            MediaSourceEventDispatcher eventDispatcher,
+            Map<String, DrmInitData> overridingDrmInitData) {
       super(allocator, drmSessionManager, eventDispatcher);
       this.overridingDrmInitData = overridingDrmInitData;
     }
 
-    public void setDrmInitData(@Nullable DrmInitData drmInitData) {
+    void setCurrentLoadingChunk(HlsMediaChunk chunk) {
+      double tolerance = (chunk.endTimeUs - chunk.startTimeUs) * MAX_TIMESTAMP_DEVIATION_PERCENTAGE;
```

Comment:
For small segments this can end up allowing quite a small tolerance. I wonder whether it's worth adding a minimum tolerance (e.g., 5 seconds) as well, and `Math.max`-ing the two? I kind of wonder whether just a fixed tolerance (e.g., 10 seconds) might be easier to reason above, since I'm not sure you'd necessarily want to allow a really large deviation for long segments either.

Reply:
I agree.   I have increased this in my own branch and testing to a larger value,  this is particularly evident for i-Frame only which can be of extremely short times.  The `Math.max()` is a great way to do this.

Reply:
See commit 4859207 changed to make max the tolerance against a min possible value.